src/GovUk.Education.ExploreEducationStatistics.Data.Services/Converters/ResultSubjectMetaViewModelJsonConverter.cs

```diff
@@ -0,0 +1,89 @@
+﻿#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.ViewModels.Meta;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace GovUk.Education.ExploreEducationStatistics.Data.Services.Converters
+{
+    /// <summary>
+    /// JsonConverter which transforms the legacy 'Locations' field of ResultSubjectMetaViewModel to 'LocationsHierarchical'.
+    ///
+    /// Intended to be temporary until this transformation is made permanent by migrating old Permalinks in blob storage (EES-2943).
+    ///
+    /// Note that this doesn't infer any hierarchy so the location attributes will continue to be single level
+    /// meaning there's no overall change to the data of Permalinks except for moving it between fields and transformation of format.
+    ///
+    /// Permalinks created prior to the Table Result Subject Metadata changes of EES-2881 (BE) and EES-2777 (FE) have a
+    /// flat 'Locations' field in their JSON serialization of type <see cref="List{ObservationalUnitMetaViewModel}"/>.
+    ///
+    /// Permalinks created afterwards, plus any created while the dedicated release toggle for that feature was turned on,
+    /// have locations in field <see cref="ResultSubjectMetaViewModel.LocationsHierarchical">ResultSubjectMetaViewModel.LocationsHierarchical</see>.
+    ///
+    /// Until old Permalinks are migrated, the translation provided by this converter ensures that consumers
+    /// are aware of legacy locations when accessing <see cref="ResultSubjectMetaViewModel.LocationsHierarchical" />.
+    /// </summary>
+    public class ResultSubjectMetaViewModelJsonConverter : JsonConverter<ResultSubjectMetaViewModel>
+    {
+        public override bool CanWrite => false;
+
+        public override ResultSubjectMetaViewModel ReadJson(JsonReader reader,
```

Comment:
Why the need for this class AND the ContractResolver out of interest?

Reply:
Oh I see, the ContractResolver uses this if it decides that it needs to?

Reply:
Yes, the ContractResolver is aware of the feature toggle.

It stops an empty Locations field getting serialized when turned on.
It stops an empty LocationsHierarchical field getting serialized when turned off.

It also applies the JsonConverter when turned on so that old Permalinks have their Locations deserialised as LocationsHierarchical.

I think once the feature toggle is removed from the codebase and it becomes permanent the ContractResolver can be removed and we will always want to apply the converter using an annotation on `TableResultViewModel` instead:

```
    [JsonConverter(typeof(ResultSubjectMetaViewModelJsonConverter))]
    public ResultSubjectMetaViewModel SubjectMeta { get; set; }
```

If we eventually migrate all permalinks then the JsonConverter can also be dropped.

Reply:
Great!  Thanks for the detailed reply btw - always appreciated :)